### PR TITLE
rustdoc: Make "rust code block is empty" and "could not parse code block" warnings a lint (`INVALID_RUST_CODEBLOCKS`)

### LINKED ISSUE #79792: rustdoc -D warnings doesn't fail for all warnings
#### I tried this code:

```bash
/// See [`foo:Foo`].
/// ```rust
fn main() {}

```

#### I expected to see this happen:
The env variable `-Dwarnings` is supposed to make rustdoc treat warnings as errors. For `export RUSTDOCFLAGS=""` the run below outputs only warnings, no errors. For `RUSTDOCFLAGS="-Dwarnings"` I would have expected both warnings to now be errors.

```
$ export RUSTDOCFLAGS="-Dwarnings"
$ cargo doc
 Documenting rustdoc-bug v0.1.0 (/tmp/bug/rustdoc-bug)
error: unresolved link to `foo:Foo`
 --> src/main.rs:1:10
  |
1 | /// See [`foo:Foo`].
  |          ^^^^^^^^^ no item named `foo:Foo` in scope
  |
  = note: `-D broken-intra-doc-links` implied by `-D warnings`
  = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

error: Rust code block is empty
 --> src/main.rs:2:5
  |
2 | /// ```rust
  |     ^^^^^^^

error: aborting due to previous error; 1 warning emitted```
```

#### Instead, this happened:
Only one of the warnings is switched to an error.

```
$ export RUSTDOCFLAGS="-Dwarnings"
$ cargo doc
 Documenting rustdoc-bug v0.1.0 (/tmp/bug/rustdoc-bug)
error: unresolved link to `foo:Foo`
 --> src/main.rs:1:10
  |
1 | /// See [`foo:Foo`].
  |          ^^^^^^^^^ no item named `foo:Foo` in scope
  |
  = note: `-D broken-intra-doc-links` implied by `-D warnings`
  = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

warning: Rust code block is empty
 --> src/main.rs:2:5
  |
2 | /// ```rust
  |     ^^^^^^^

error: aborting due to previous error; 1 warning emitted```
```

This is also problematic because for warnings the exit code is still `0`, so a CI setup might not detect the failure:

```
$ cat src/main.rs 
/// ```rust
fn main() {}
$ RUSTDOCFLAGS="-Dwarnings" cargo doc
warning: Rust code block is empty
 --> src/main.rs:1:5
  |
1 | /// ```rust
  |     ^^^^^^^

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
$ echo $?
0
```

### Meta

`rustc --version --verbose`:
```
rustc 1.50.0-nightly (1c389ffef 2020-11-24)
```

### PR DESCRIPTION
**Pull request by jyn514**
Fixes https://github.com/rust-lang/rust/issues/79792. This already went through FCP in https://github.com/rust-lang/rust/pull/79816, so it only needs final review.

This is mostly a rebase of https://github.com/rust-lang/rust/pull/79816 - thank you @poliorcetics for doing most of the work!

### CODE REVIEW

File: src/doc/rustdoc/src/lints.md [hunk 1]
```diff
@@ -294,6 +294,43 @@ warning: unclosed HTML tag `h1`
...
+4 | | /// ```
+| |_______^
+|
+= note: `#[warn(rustdoc::invalid_rust_codeblock)]` on by default
+
+warning: Rust code block is empty
+--> src/lib.rs:8:5
+|
+8 | /// ```rust
+|     ^^^^^^^
```
Comment (Smittyvb) [inline src/doc/rustdoc/src/lints.md:331]: The carets seem a bit off here?
```suggestion
|       ^^^^^^^
```

Comment (jyn514) [reply src/doc/rustdoc/src/lints.md:331]: I think this is just from me mangling copy/paste, I updated the docs.

**Comment by CraftSpider** [comment]
I don't see a test for the actual empty block case, could you add one? Otherwise looks generally good, given that I'm not super familiar with the lint system.

**Comment by jyn514** [comment]
> I don't see a test for the actual empty block case, could you add one? 

It already exists: https://github.com/rust-lang/rust/blob/7f4d7327da6dbec5958d144a12e2e4709ebcbfc1/src/test/rustdoc-ui/invalid-syntax.stderr#L105-L111

Err ... actually I don't think there was ever an FCP in https://github.com/rust-lang/rust/pull/79816. Also the lint name should be plural, not singular. I'll start FCP once I rename it.

**Comment by jyn514** [comment]
@rfcbot fcp merge

This adds a new lint `INVALID_RUST_CODEBLOCKS`. These warnings were already being emitting, but as a hard warning and not a lint, so they couldn't be allowed or denied.

I'm open to renaming the lint, but I do think changing these from a warning to a lint is the right decision.

**Comment by jyn514** [comment]
@rfcbot fcp merge

This adds a new lint `INVALID_RUST_CODEBLOCKS`. These warnings were already being emitting, but as a hard warning and not a lint, so they couldn't be allowed or denied.

I'm open to renaming the lint, but I do think changing these from a warning to a lint is the right decision.

File: src/test/rustdoc-ui/ignore-block-help.stderr [hunk 2]
```diff
@@ -7,11 +7,13 @@ LL | | /// let heart = '❤️';
...
-help: `ignore` code blocks require valid Rust code for syntax highlighting. Mark blocks that do not contain Rust code as text
+   = note: `#[warn(rustdoc::invalid_rust_codeblocks)]` on by default
+help: `ignore` code blocks require valid Rust code for syntax highlighting; mark blocks that do not contain Rust code as text: ```text
+  --> $DIR/ignore-block-help.rs:3:5
    |
-LL | /// ```text,ignore (to-prevent-tidy-error)
-   |     ^^^^^^^^
+LL | /// ```ignore (to-prevent-tidy-error)
+   |     ^^^
+   = note: error from rustc: character literal may only contain one codepoint
```
Comment (Nemo157) [inline src/test/rustdoc-ui/ignore-block-help.stderr:16]: Is there an existing issue about this not pointing to the span that contains the error? Would be nice to fix in the future.

File: src/doc/rustdoc/src/lints.md [hunk 3]
```diff
@@ -294,6 +294,43 @@ warning: unclosed HTML tag `h1`
...
+```rust
+/// Empty code blocks (with and without the `rust` marker):
+///
+/// ```rust
+/// ```
+///
+/// Unclosed code blocks (with and without the `rust` marker):
+///
+/// ```rust
+fn main() {}
```
Comment (camelid) [inline src/doc/rustdoc/src/lints.md:311]: Might be worth giving an example of invalid syntax too.

File: src/test/rustdoc-ui/ignore-block-help.stderr (hunk 2 above)
Comment (camelid) [reply src/test/rustdoc-ui/ignore-block-help.stderr:16]: Why was this behavior changed in this PR?

**Comment by camelid** [comment]
@rfcbot reviewed

I agree with making this a lint, but why was https://github.com/rust-lang/rust/pull/84587#discussion_r624580967 changed in this PR? I would prefer it to be a separate change.

File: src/test/rustdoc-ui/ignore-block-help.stderr (hunk 2 above)
Comment (Nemo157) [reply src/test/rustdoc-ui/ignore-block-help.stderr:16]: The `note` just moved below the `help`, it wasn't attached to the correct span before either.

File: compiler/rustc_mir/src/borrow_check/region_infer/mod.rs [hunk 4]
```diff
@@ -1241,7 +1241,7 @@ impl<'tcx> RegionInferenceContext<'tcx> {
     /// it. However, it works pretty well in practice. In particular,
     /// this is needed to deal with projection outlives bounds like
     ///
-    /// ```ignore (internal compiler representation so lifetime syntax is invalid)
+    /// ```text (internal compiler representation so lifetime syntax is invalid)
```
Comment (GuillaumeGomez) [inline compiler/rustc_mir/src/borrow_check/region_infer/mod.rs:1244]: There is no need for the explanation insde the paren, it's only required with `ignore`.

File: src/librustdoc/passes/check_code_block_syntax.rs [hunk 5]
```diff
@@ -47,63 +48,86 @@ impl<'a, 'tcx> SyntaxChecker<'a, 'tcx> {
...
-                let mut diag = self.cx.sess().struct_span_warn(sp, warning_message);
-
-                if code_block.syntax.is_none() && code_block.is_fenced {
-                    let sp = sp.from_inner(InnerSpan::new(0, 3));
-                    diag.span_suggestion(
-                        sp,
-                        "mark blocks that do not contain Rust code as text",
-                        String::from("```text"),
-                        Applicability::MachineApplicable,
+        if !(buffer.has_errors || is_empty) {
```
Comment (GuillaumeGomez) [inline src/librustdoc/passes/check_code_block_syntax.rs:51]: Not a big fan of paren conditions. :p

```suggestion
        if !buffer.has_errors && !is_empty {
```

File: src/librustdoc/passes/check_code_block_syntax.rs [hunk 6]
```diff
@@ -47,63 +48,86 @@ impl<'a, 'tcx> SyntaxChecker<'a, 'tcx> {
...
+            // No errors in a non-empty program.
+            return;
+        }
+
+        let local_id = match item.def_id.as_local() {
```
Comment (GuillaumeGomez) [inline src/librustdoc/passes/check_code_block_syntax.rs:56]: Just wondering here: wouldn't using `?` be simpler than matching?

Comment (jyn514) [reply src/librustdoc/passes/check_code_block_syntax.rs:56]: `?` doesn't work if the function doesn't return a value.

File: src/test/rustdoc-ui/ignore-block-help.stderr (hunk 2 above)
Comment (jyn514) [reply src/test/rustdoc-ui/ignore-block-help.stderr:16]: The existing issue is https://github.com/rust-lang/rust/issues/67857.

File: src/doc/rustdoc/src/lints.md (hunk 3 above)
Comment (jyn514) [reply src/doc/rustdoc/src/lints.md]: Sure thing, done.

File: src/test/rustdoc-ui/ignore-block-help.stderr (hunk 2 above)
Comment (camelid) [reply src/test/rustdoc-ui/ignore-block-help.stderr:16]: > it wasn't attached to the correct span before either.

Ah, I see 👍

**Comment by jyn514** [comment]
@CraftSpider this is just waiting on your review :)

**Review by CraftSpider — APPROVED** [review]

**Comment by CraftSpider** [comment]
Shall I `r+`? I'm not 100% on the FCP finalization protocol :P

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Since the CFP is complete, you can r+. (I let you do the command ;) )

**Comment by GuillaumeGomez** [comment]
I think it'll need to be rebased:

```
    Checking tracing-subscriber v0.2.16
    Checking tracing-tree v0.1.9
    Checking rustdoc-json-types v0.1.0 (/checkout/src/rustdoc-json-types)
    Checking rustdoc v0.0.0 (/checkout/src/librustdoc)
error[E0599]: no method named `as_local` found for enum `types::FakeDefId` in the current scope
  --> src/librustdoc/passes/check_code_block_syntax.rs:56:42
   |
56 |         let local_id = match item.def_id.as_local() {
   |                                          ^^^^^^^^ help: there is an associated function with a similar name: `is_local`
  ::: src/librustdoc/clean/types.rs:54:1
   |
   |
54 | crate enum FakeDefId {
   | -------------------- method `as_local` not found for this
error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
error: could not compile `rustdoc`
```

**Pull request merged**